### ISSUE TITLE
Custom environment, design question

### ISSUE THREAD

**Issue by sereysethy** [rlworkgroup/garage#1376]
Hello,

I want to design a custom environment, I have two questions:

**Question 1**: my observation is composed of one (or more) discrete value(s) and several real values. What should I use? I think of two possibilities.

1. Use `Box` of `n` dimension, in this case a discrete value can also be represented as a real number.
2. Use `Dict` which consists of one discrete variable and one box variable.

```python
self.observation_space = akro.Dict({"dis_var": akro.Discrete(1),
                                            "real_var": akro.Box(4)})
```
**Question 2**: should I use `akro` or `gym.spaces` to define my custom environment?

**Comment by krzentner** [comment]
Hi Sereysethy,

Thanks for taking an interest in garage.
About the first question, the current state of garage makes this complicated. I definitely think a `Dict` space as you describe is the right way to go overall. However, the typical benchmarks we use for testing (mostly `gym`) only use `Box` observations. Because of that, most of garage is not yet tested against `Dict` observations. In particular only `garage.tf.algos.DDPG` and `garage.tf.algos.TD3` are regularly tested against `Dict` observations (and `DDPG` has a performance regression at the moment).

If you're willing to encounter a few bugs (which I'd be happy to help fix), I'd recommend using a `Dict` space, otherwise use a `Box`.

About the second question, I (as the primary `akro` maintainer), see it as a strict improvement on `gym.spaces`. All of the `akro` spaces inherit related `gym` spaces. Furthermore, the `flatten` functionality allows you the flexibility of using non-`Box` spaces, and turning them into `Box` style observations if needed.

Best,

K.R.

**Comment by sereysethy** [comment]
Hi KR,

Thank you for your reply. I will go with `Dict`.

I have another question, I think this might be a silly one. If I use `Box`, I can have an array of `n-dimension`, but I can also use `Dict` of `n-element`, the benefit of using `Dict` is that I can give a name of each of my feature. Which one do you think it is more convenient?

Cheers,
Sethy

**Comment by sereysethy** [comment]
Another question, when I create a new environment, should I inherit `Gym.env` or `GarageEnv`?

Env has to be normalised? Or it depends on algorithm? I am a bit confused.

**Comment by ryanjulian** [comment]
@sereysethy you could use dict of n-elements if you really wanted to (that is akin to turning a list into a dict like `{i: e for i, e in enumerate(list)}`). Most of the time people don't do this in Python because dictionaries are more expensive than lists, but in all honestly it probably doesn't matter. Which is more convenient is up to you, but reiterating what @krzentner said, Dict spaces are not very well tested in garage right now, and you *will* encounter bugs when using them.

1. You can inherit either or none. For portability, inherit `gym.Env`. As long as your environment implements all of the functions on the `gym.Env` interface, it should work.
2. Whether you normalize the environment is an algorithmic or implementation choice. Where you do it is arbitrary. In general, RL environments do not need to be normalized. Many RL algorithms perform better with normalized rewards and/or observations.
3. Sampling is handled by the `garage.sampler` package, and I recommend you read the interface in `src/garage/sampler/sampler.py`, `src/garage/sampler/worker_factory.py`, and `src/garage/sampler/worker.py`. You can set your chosen sampler and worker classes in a call to `LocalRunner.setup`. Based on what you are saying, it seems like you don't need to change any of this, and can instead just seek accept user input on a call to `gym.Env.step`. Note that, by default, sampling in garage is parallelized, so you should probably override your `sampler_class` to `LocalSampler` to avoid having many environment instances.
4. I want to reiterate that `LinearFeatureBaseline` is not an algorithm, but a component of an algorithm. In particular, it helps garage algorithms implement [Generalized Advantage Estimation](https://arxiv.org/abs/1506.02438). Are you seeking to implement value iteration (Sutton & Barto, Ch 4.4?)? In that case, garage has not implementation of this algorithm because it's not in widespread use in the RL research community.

**Comment by sereysethy** [comment]
Hi Ryan,

Thank you for your reply. I still have some more follow-up questions, sorry to ask so many questions.

1. I will got with simple Box but then what to do with my discrete values? The discrete values represent user inputs. So I will just represent it as `np.float32`?
2. At this point, I will not bother with this just yet. I will come back later for this.
3. Can you be more specific for sampling? How `sampler_class` can be overridden in `LocalSampler`? How `LocalSampler` is called in the `LocalRunner`?
4. I want to implement simple algorithms like SARSA or Q-Learning using a function approximator like Linear function and later MLP, before trying DQN or PG. The idea is to learn how garage works and at the same time use it as my baseline.

What I am doing now is: 1) creating a custom environment, 2) implement simple algorithm to test the idea.

**Comment by ryanjulian** [comment]
1. Use the `akro.Discrete` or `gym.spaces.Discrete` space for discrete-valued quantities. `Box` spaces are continuously-valued.
3. Unfortunately I think it will just take too many words for me to explain this point right now, but we will note it as a piece of high-level documentation which we need to write. I believe reading the code for some algorithms, the sampler, `LocalRunner` will be enlightening on this point. I've included a list of links at the bottom to get you started.
4. Garage would be good for doing this. If you want to use a linear function approximator for your Q-function or policy, you can implement this by creating a custom Q-function or policy class.  Alternatively, if you set the `hidden_nonlinearity` and `output_nonlinearity` parameters of  `garage.torch.policies.DeterministicMLPPolicy`, you would get a purely-linear function approximator with your chosen numbers of layers and units (and similarly for other neural network classes in torch and TF). DQN is the algorithm in garage which most closely-approximates pure Q-learning (it simply replaces a tabular Q-function with a function approximator). Currently, DQN is implemented only in `garage.tf`.

If you are implementing simple, classic algorithms like SARSA or tabular Q-learning, you might consider working in the `garage.np` tree, which is components which don't require neural networks. The advantage of doing this is that it will allow you to avoid the complexities of neural network training, while the disadvantage is that there are far fewer examples and pre-made components in `garage.np`, due to its relative lack of popularity.

Code snippets which show use of the sampling API:
Overriding the default sampler class: https://github.com/rlworkgroup/garage/blob/master/examples/tf/trpo_swimmer_ray_sampler.py
Overriding the default worker class:
https://github.com/rlworkgroup/garage/blob/master/examples/tf/rl2_ppo_halfcheetah.py
Shows the interaction between the runner, sampler, and algorithm: https://github.com/rlworkgroup/garage/blob/master/src/garage/np/algos/batch_polopt.py#L76
Code for LocalRunner: group/garage/blob/master/src/garage/experiment/local_runner.py#L1

**Comment by sereysethy** [comment]
1. As I wrote in the beginning, my observation consists of discrete and continuous values, that is why I wanted to use `Dict`, but then you said `Dict` is not yet well tested with all the algo. So what am I supposed to do?
2. For sampler, I will take a look at it.
3. I will take a look at `garage.torch.policies.DeterministicMLPPolicy`.

Tabular methods will not be useful in my case due to my feature but I might also need them for comparison.

**Comment by ryanjulian** [comment]
1. You can either accept the instability of `Dict` and work to add the features you need for them (alternatively, report issues and wait for us to fix them), or you can cast all of your observations floats, concatenate them, and use a Box space. If you plan on combining your real- and discrete-valued observations in the same function approximator, you will have to cast all of the discrete values into floats anyway.

**Comment by sereysethy** [comment]
Thank you. I will go with Box.
